feat(daffio): add anchor links to markdown titles h2,h3,h4

ali-toghiani opened the pull request:
https://github.com/graycoreio/daffodil/issues/3973

## PR Checklist

- [x] Commit message follows our [contributing guidelines](https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit)
- [ ] Tests added/updated (for bug fixes/features)
- [ ] Documentation added/updated (for bug fixes/features)

## PR Type
<!-- Select the relevant type(s) -->

- [ ] Bug fix
- [x] Feature
- [ ] Style update
- [ ] Refactor
- [ ] Test
- [ ] Build
- [ ] CI
- [ ] Docs
- [ ] Performance
- [ ] Other (please describe)

## Current behavior
<!-- Describe the current behavior or link to relevant issue -->

Fixes: #3973

## New behavior
<!-- Describe the changes -->
This solution uses custom-elements instead of pure inline html, css & js
## Breaking change?

- [ ] Yes
- [x] No

<!-- If yes, describe impact and migration path -->

## Additional context
<!-- Any other relevant information -->

ali-toghiani committed:
feat(daffio): add anchor links to markdown titles h2,h3,h4

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
feat(daffio): add anchor links to markdown titles h2,h3,h4

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove useless comment

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove useless comment

https://github.com/graycoreio/daffodil/issues/3973

griest024 reviewed (changes requested):
This should be implemented as an Angular component using `@angular/elements`

xelaint commented:
Can you clarify the name of this PR or #4023? They're the same, and it's confusing what each PR is meant to accomplish.

damienwebdev commented:
@xelaint see the discussion in #3973 - @griest and I discussed an alternative implementation and @ali-toghiani was kind enough to do both implementations!

ali-toghiani committed:
refactor(daffio): replace cd-heading.cd.ts with heading.component

remove the ce-heading.ce.ts custom-element
created an angular heading component, which is imported to the daffio project using createCustomElement from '@angular/elements'

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): replace cd-heading.cd.ts with heading.component

remove the ce-heading.ce.ts custom-element
created an angular heading component, which is imported to the daffio project using createCustomElement from '@angular/elements'

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani commented:
@griest024 I removed the ce-heading.ce.ts custom-element and instead created an angular component called heading and imported it to the project using the createCustomElement, is this what you meant ?

griest024 commented on apps/daffio/src/app/app.module.ts:
```diff
@@ -70,4 +76,16 @@ import { TemplateModule } from './core/template/template.module';
+    @Inject(PLATFORM_ID) private platformId: object,
+    private injector: Injector,
+  ) {
+    if (isPlatformBrowser(this.platformId)) {
```
What are you plans for supporting SSR?

ali-toghiani replied:
well if we use the new customElemenetsProvider, it skips registering the component on the server, so it wont throw any errors there,
then angular processes this component and considers it as unknown html element, which if we add the `CUSTOM_ELEMENTS_SCHEMA` to the providers array, I suppose it will be rendered ?
I really tried running the `dev:ssr` command but it just keeps getting stuck on the Building... phase.

griest024 replied:
hmm if daffio is already built (`npx nx run daffio:build`) you can just run the SSR server through the vscode debug config in `launch.json`.

griest024 commented on apps/daffio/src/app/docs/components/heading/heading.component.html:
```diff
@@ -0,0 +1,22 @@
+	[attr.title]="'Copy link to this section'"
+	(click)="onAnchorClick($event)"
+>
+	<svg
+		class="daffio-docs-heading-link__icon"
+		xmlns="http://www.w3.org/2000/svg"
+		viewBox="0 0 512 512"
+		width="12"
+		height="12"
+		aria-hidden="true"
+	>
+		<path
+			[attr.fill]="iconColor"
+			d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-2 ...
+		></path>
+	</svg>
```
go ahead and use an icon from font awesome: https://fontawesome.com/icons/link?f=classic&s=solid for consistency. See other components in daffio for examples.

ali-toghiani replied:
the plain svg was used in the old pure custom-element component which should've been refactored, it's replaced now
```html
<fa-icon [icon]="faLink" size="sm" class="daffio-ce-heading__icon"></fa-icon>
```

griest024 commented on apps/daffio/src/app/docs/components/heading/heading.component.ts:
```diff
@@ -0,0 +1,40 @@
+  constructor(private el: ElementRef) {}
+
+  ngAfterViewInit() {
+    const heading = this.el.nativeElement.querySelector('h1, h2, h3, h4, h5, h6');
+    if (heading) {
+      heading.classList.add('daffio-ce-heading');
+    }
```
this is not hydration safe. What are you trying to accomplish here? This class doesn't have any styles associated with it.

ali-toghiani replied:
The purpose of this segment was to unify ce-heading and other headers like heading-link styles, which was not necessary, I removed this segment, thanks for you review

griest024 commented on apps/daffio/src/app/docs/components/heading/heading.component.ts:
```diff
@@ -0,0 +1,40 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class MarkdownHeadingComponent implements AfterViewInit {
+  @Input() size = 'h1';
```
This input needs stricter typing

ali-toghiani replied:
I defined a HeadingSize type like this in the heading.component.ts file, let me know if there are any issues with it
```ts
export type HeadingSize = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
```

griest024 commented on apps/daffio/src/app/app.module.ts:
```diff
@@ -70,4 +76,16 @@ import { TemplateModule } from './core/template/template.module';
+    private injector: Injector,
+  ) {
+    if (isPlatformBrowser(this.platformId)) {
+      if (!customElements.get('daffio-ce-heading')) {
+        const heading = createCustomElement(MarkdownHeadingComponent, { injector: this.injector });
+        customElements.define('daffio-ce-heading', heading);
+      }
```
please move this to an `AppInitializer` provider.

ali-toghiani replied:
I created a provider for all angular-elements components, which is located at
`daffio/src/app/core/providers/custom-elements-provider.ts`
I'm not sure about the location, initially I created it under `docs/` but after refactoring the provider to provide multiple custom-elements, I thought maybe it's better to use a more general location.
I also noticed that the Class Name doesn't follow the Daffio naming conventions and fixed that one too.

xelaint commented on apps/daffio/src/app/docs/components/heading/heading.component.html:
```diff
@@ -0,0 +1,22 @@
+<ng-content></ng-content>
+<a
+	class="daffio-docs-heading-link daffio-heading-link"
```
Don't think both classes are needed, and the class name should reflect component selector name. `daffio-ce-heading`

ali-toghiani replied:
css class name converter, the extra class is removed 👍

xelaint commented on apps/daffio/src/app/docs/components/heading/heading.component.scss:
```diff
@@ -0,0 +1,49 @@
+	align-items: baseline;
+	position: relative;
+
+	::ng-deep {
```
`::ng-deep` is not necessary here.

ali-toghiani replied:
removed 👍

xelaint commented on apps/daffio/src/app/docs/components/heading/heading.component.scss:
```diff
@@ -0,0 +1,49 @@
+
+			&.copied {
+				&__icon {
+					color: #28a745;
```
color should be in a separate `heading-theme.scss` file so that it covers both dark and light mode. See other components with a theme file for examples.

ali-toghiani replied:
nice catch, created heading-theme.scss, also fixed copied green color not working, 
in light mode green-50 and in dark mode green-20 is used, I checked them & they looked good enough to me, let me know if there is a specific design for that

xelaint commented on apps/daffio/src/app/docs/components/heading/heading.component.html:
```diff
@@ -0,0 +1,22 @@
+	class="daffio-docs-heading-link daffio-heading-link"
+	[href]="'#' + slug"
+	[attr.aria-label]="'Link to this section'"
+	[attr.title]="'Copy link to this section'"
```
title is not needed.

xelaint commented on apps/daffio/src/app/docs/components/heading/heading.component.html:
```diff
@@ -0,0 +1,22 @@
+<a
+	class="daffio-docs-heading-link daffio-heading-link"
+	[href]="'#' + slug"
+	[attr.aria-label]="'Link to this section'"
```
Should just be `aria-label="Copy link to section"`

ali-toghiani committed:
refactor(daffio): replace plain svg with fa-link icon from font awesome

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): replace plain svg with fa-link icon from font awesome

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): move custom-elements registration to an AppInitializer provider.

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): move custom-elements registration to an AppInitializer provider.

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): implement stricter typing for the size input

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): implement stricter typing for the size input

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): use colors from newly created heading-theme file

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): use colors from newly created heading-theme file

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove title from anchor tag

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove title from anchor tag

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): fix aria-label text and binding

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): fix aria-label text and binding

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove unused variable

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
style(daffio): remove unused variable

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): rename heading component's class

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani committed:
refactor(daffio): rename heading component's class

https://github.com/graycoreio/daffodil/issues/3973

griest024 commented on apps/daffio/src/app/docs/components/heading/heading.component.html:
```diff
@@ -0,0 +1,9 @@
+<ng-content></ng-content>
+<a
+	class="daffio-ce-heading"
+	[href]="'#' + slug"
```
this doesn't seem to be properly relative on design component pages, possibly because they're inside a tab (which uses the `tab` query param)

griest024 commented on apps/daffio/src/app/core/providers/custom-elements-provider.ts:
```diff
@@ -0,0 +1,53 @@
+export const provideDaffCustomElements = () => makeEnvironmentProviders([
+  provideAppInitializer(() => {
+    const platformId = inject(PLATFORM_ID);
+    if (isPlatformBrowser(platformId)) {
```
Not rendering this in SSR shouldn't be an issue for SEO content but it could cause hydration issues, though it will be inside an `innerHtml` so it actually might not matter. It doesn't seem to cause issues at the moment so I guess it doesn't need to block this PR.

griest024 reviewed (changes requested):
Clicking the icon in the design component API tab doesn't copy the link to the clipboard. Its not part of the custom element but it is in scope of the issue and needs to be implemented in the daffio heading component.

The design of adjacent headings seems to be off. This is likely a daff article issue since the heading is wrapped in an extra component.
This PR:
<img width="488" height="160" alt="Screenshot_2025-10-06_09-41-31" src="https://github.com/user-attachments/assets/0b2a70ca-3a47-4114-a5b5-e5587ba3c539" />
What it should look like (ignore dark vs light theme):
<img width="341" height="225" alt="Screenshot_2025-10-06_09-41-42" src="https://github.com/user-attachments/assets/0e6aa28f-8de1-49a4-961b-6e4741a9afa7" />
@xelaint do you have advice for an easy fix?

xelaint commented:
@ali-toghiani Can you rebase this branch from develop so it has the latest changes? I need those changes in order to review this PR. Thanks!

ali-toghiani committed:
Merge branch 'ce-heading-anchor' of https://github.com/ali-toghiani/daffodil into ce-heading-anchor
merge upstream develop into ce-heading-anchor

ali-toghiani committed:
feat(daffio): create heading component used in docs md generation

https://github.com/graycoreio/daffodil/issues/3973

ali-toghiani commented:
@xelaint some issues happened during rebase, had to force-push it, anyway it's ready to be reviewed

griest024 commented:
> @xelaint some issues happened during rebase, had to force-push it, anyway it's ready to be reviewed

a rebase will (usually) always require a force push FWIW

ali-toghiani committed:
refactor(daffio): fix copied css class linting failure

ali-toghiani committed:
Merge branch 'ce-heading-anchor' of https://github.com/ali-toghiani/daffodil into ce-heading-anchor

xelaint reviewed (changes requested):
The structure needs to be:

```
<h2>
  <a href="">
    <!--heading content-->
    <!--link icon-->
  </a>
</h2>
```

ali-toghiani committed:
refactor(daffio): move heading text inside the anchor tag

<h2>
  <a href="">
    <!--heading content-->
    <!--link icon-->
  </a>
</h2>

ali-toghiani commented:
I know that I could use Renderer2 for https://github.com/graycoreio/daffodil/pull/4027/commits/6d6a029111efcb1007ae925f75bc868fc4292820
But considering SSR compatibility, I think the static method works better

ali-toghiani committed:
style(daffio): remove !important from styles

xelaint commented:
> I know that I could use Renderer2 for [6d6a029](https://github.com/graycoreio/daffodil/commit/6d6a029111efcb1007ae925f75bc868fc4292820) But considering SSR compatibility, I think the static method works better

@ali-toghiani The updates do not work. This is what the page looks like. The content is not being rendered:

<img width="919" height="786" alt="Screen Shot 2025-10-09 at 4 19 22 PM" src="https://github.com/user-attachments/assets/58d4b64b-b93c-4aa0-a27f-25062f7a6b79" />

ali-toghiani commented:
@xelaint Did you build dgeni ? 
`nx run @daffodil/tools-dgeni:build`

xelaint commented:
@ali-toghiani Yes, I did

ali-toghiani committed:
fix(tools-dgeni): fix component docs not showing

the reason behind the issue was using the self-closing tag sytanx instead of the paired tags.

ali-toghiani commented:
@xelaint You're right, It was caused by using self-closing syntax for the daffio-ce-heading tag in the markdown.ts file,
https://github.com/graycoreio/daffodil/pull/4027/commits/4369bec9f7976f6ecf9ec5c497a090033706066f
The issue is fixed now

xelaint commented:
@ali-toghiani It seems to be very buggy. The headings appear on initial render, but navigating to the API tab and then back to the Usage tab will make the headings disappear. The headings do reappear on page refresh.

griest024 commented:
This PR appears to be abandoned, closing
